### ISSUE TITLE
Karpenter is nominating pods to machine that does not match its requirements

### ISSUE THREAD

**Issue by myaser** [aws/karpenter#4438]
### Description

**Observed Behavior**:
launching a pod with 
```
      nodeSelector:
        karpenter.k8s.aws/instance-local-nvme: "50"
```
I read in the events:
1. kube_scheduler failed to schedule the pod
```
0/9 nodes are available: 1 node(s) had untolerated taint {dedicated: skipper-ingress-redis}, 1 node(s) had untolerated taint {node.kubernetes.io/role: master}, 2 node(s) had untolerated taint {dedicated: skipper-ingress}, 5 node(s) didn't match Pod's node affinity/selector. preemption: 0/9 nodes are available: 9 Preemption is not helpful for scheduling.
```
2. karpenter nominated the pod to an existing machine/node
```
Pod should schedule on: machine/default-karpenter-f72cc, node/ip-10-149-71-113.eu-central-1.compute.internal
```
this node does not have the right label
```
NAME                                             STATUS   ROLES    AGE   VERSION    LABELS
ip-10-149-71-113.eu-central-1.compute.internal   Ready    worker   16h   v1.24.16   aws.amazon.com/spot=true,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=r5n.xlarge,beta.kubernetes.io/os=linux,cluster-lifecycle-controller.zalan.do/replacement-strategy=none,failure-domain.beta.kubernetes.io/region=eu-central-1,failure-domain.beta.kubernetes.io/zone=eu-central-1a,karpenter.k8s.aws/instance-category=r,karpenter.k8s.aws/instance-cpu=4,karpenter.k8s.aws/instance-encryption-in-transit-supported=true,karpenter.k8s.aws/instance-family=r5n,karpenter.k8s.aws/instance-generation=5,karpenter.k8s.aws/instance-hypervisor=nitro,karpenter.k8s.aws/instance-memory=32768,karpenter.k8s.aws/instance-network-bandwidth=4100,karpenter.k8s.aws/instance-pods=58,karpenter.k8s.aws/instance-size=xlarge,karpenter.sh/capacity-type=spot,karpenter.sh/initialized=true,karpenter.sh/provisioner-name=default-karpenter,karpenter.sh/registered=true,kubernetes.io/arch=amd64,kubernetes.io/hostname=ip-10-149-71-113,kubernetes.io/os=linux,kubernetes.io/role=worker,lifecycle-status=ready,node.kubernetes.io/distro=ubuntu,node.kubernetes.io/instance-type=r5n.xlarge,node.kubernetes.io/node-pool=default-karpenter,node.kubernetes.io/profile=worker-karpenter,node.kubernetes.io/role=worker,topology.ebs.csi.aws.com/zone=eu-central-1a,topology.kubernetes.io/region=eu-central-1,topology.kubernetes.io/zone=eu-central-1a
```
```
> zkubectl describe machine default-karpenter-f72cc
...
Labels:       cluster-lifecycle-controller.zalan.do/replacement-strategy=none
              karpenter.k8s.aws/instance-category=r
              karpenter.k8s.aws/instance-cpu=4
              karpenter.k8s.aws/instance-encryption-in-transit-supported=true
              karpenter.k8s.aws/instance-family=r5n
              karpenter.k8s.aws/instance-generation=5
              karpenter.k8s.aws/instance-hypervisor=nitro
              karpenter.k8s.aws/instance-memory=32768
              karpenter.k8s.aws/instance-network-bandwidth=4100
              karpenter.k8s.aws/instance-pods=58
              karpenter.k8s.aws/instance-size=xlarge
              karpenter.sh/capacity-type=spot
              karpenter.sh/provisioner-name=default-karpenter
              kubernetes.io/arch=amd64
              kubernetes.io/os=linux
              lifecycle-status=ready
              node.kubernetes.io/instance-type=r5n.xlarge
              node.kubernetes.io/node-pool=default-karpenter
              node.kubernetes.io/profile=worker-karpenter
              node.kubernetes.io/role=worker
              topology.kubernetes.io/region=eu-central-1
              topology.kubernetes.io/zone=eu-central-1a
...
Events:
  Type    Reason                 Age                 From       Message
  ----    ------                 ----                ----       -------
  Normal  Unconsolidatable       51m                 karpenter  not all pods would schedule, visibility/zmon-appliance-worker-7bbb86d9dc-wzw67 => would schedule against a non-initialized node default-karpenter-scs5j visibility/zmon-appliance-worker-7bbb86d9dc-k52g7 => would schedule against a non-initialized node default-karpenter-scs5j visibility/zmon-appliance-worker-7bbb86d9dc-58z65 => would schedule against a non-initialized node default-karpenter-scs5j default/testing-nginx-665669cfc5-8wqrp => would schedule against a non-initialized node default-karpenter-scs5j
  Normal  DeprovisioningBlocked  30m (x8 over 53m)   karpenter  Cannot deprovision machine due to pdb visibility/zmon-appliance-worker prevents pod evictions
  Normal  Unconsolidatable       14m (x2 over 29m)   karpenter  can't remove without creating 2 candidates
  Normal  DeprovisioningBlocked  14s (x27 over 92m)  karpenter  Cannot deprovision machine due to machine is nominated
```

karpenter logs were silent about it, even with the debug level

**Expected Behavior**:
expect karpenter to nominate the pod to a valid node, if it can not schedule the pod, it should be visible in the pod events and/or karpenter logs

**Reproduction Steps** (Please include YAML):
deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: testing-nginx
  name: testing-nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: testing-nginx
  template:
    metadata:
      labels:
        app: testing-nginx
    spec:
      nodeSelector:
        karpenter.k8s.aws/instance-local-nvme: "50"
      containers:
      - image: nginx
        name: nginx
        resources: 
          limits:
            cpu: 200m
            memory: 50Mi
          requests:
            cpu: 200m
            memory: 50Mi
```
provisioner
```
- apiVersion: karpenter.sh/v1alpha5
  kind: Provisioner
  metadata:
    name: default-karpenter
  spec:
    consolidation:
      enabled: true
    labels:
      cluster-lifecycle-controller.zalan.do/replacement-strategy: none
      lifecycle-status: ready
      node.kubernetes.io/node-pool: default-karpenter
      node.kubernetes.io/profile: worker-karpenter
      node.kubernetes.io/role: worker
    providerRef:
      name: default-karpenter
    requirements:
    - key: karpenter.k8s.aws/instance-family
      operator: In
      values:
      - c5
      - m5
      - r5
      - c5d
      - m5d
      - r5d
      - c5n
      - m5n
      - r5n
      - c6i
      - m6i
      - r6i
      - c6id
      - m6id
      - r6id
      - c6in
      - m6in
      - r6in
    - key: karpenter.k8s.aws/instance-size
      operator: NotIn
      values:
      - metal
    - key: karpenter.sh/capacity-type
      operator: In
      values:
      - spot
      - on-demand
    - key: topology.kubernetes.io/zone
      operator: In
      values:
      - eu-central-1a
      - eu-central-1b
      - eu-central-1c
    - key: kubernetes.io/os
      operator: In
      values:
      - linux
    startupTaints:
    - effect: NoSchedule
      key: zalando.org/node-not-ready
```

**Versions**:
- Chart Version: 
Not using Helm, karpenter version is v0.29.2
- Kubernetes Version (`kubectl version`): 
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.16", GitCommit:"51e33fadff13065ae5518db94e84598293965939", GitTreeState:"clean", BuildDate:"2023-07-19T12:19:24Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/arm64"}

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments; they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by bwagner5** [comment]
I haven't been able to reproduce this. 

I did a test by starting up a c6i.8xlarge (so plenty of space) and then applying your test-nginx deployment. Karpenter correctly spins up a new node with local nvme to match the node selector.

**Comment by bwagner5** [comment]
Oh actually I can reproduce this. I was using HEAD rather than v0.29.2. It appears this is fixed at HEAD.  More updates to come on what happened.

**Comment by bwagner5** [comment]
So looks like this was resolved in this PR: https://github.com/aws/karpenter-core/pull/432 and will be released next week with v0.30.0
